pdf/model/appender.go

```diff
@@ -0,0 +1,481 @@
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/unidoc/unidoc/common"
+	"github.com/unidoc/unidoc/pdf/core"
+)
+
+// PdfAppender appends a new Pdf content to an existing Pdf document.
+type PdfAppender struct {
+	rs       io.ReadSeeker
+	parser   *core.PdfParser
+	roReader *PdfReader
+	Reader   *PdfReader
+	pages    []*PdfPage
+	acroForm *PdfAcroForm
+
+	xrefs          core.XrefTable
+	greatestObjNum int
+
+	newObjects   []core.PdfObject
+	hasNewObject map[core.PdfObject]struct{}
+}
+
+func getPageResources(p *PdfPage) map[core.PdfObjectName]core.PdfObject {
+	resources := make(map[core.PdfObjectName]core.PdfObject)
+	if p.Resources == nil {
+		return resources
+	}
+	if p.Resources.Font != nil {
+		if dict, found := core.GetDict(p.Resources.Font); found {
+			for _, key := range dict.Keys() {
+				resources[key] = dict.Get(key)
+			}
+		}
+	}
+	if p.Resources.ExtGState != nil {
+		if dict, found := core.GetDict(p.Resources.ExtGState); found {
+			for _, key := range dict.Keys() {
+				resources[key] = dict.Get(key)
+			}
+		}
+	}
+	if p.Resources.XObject != nil {
+		if dict, found := core.GetDict(p.Resources.XObject); found {
+			for _, key := range dict.Keys() {
+				resources[key] = dict.Get(key)
+			}
+		}
+	}
+	if p.Resources.Pattern != nil {
+		if dict, found := core.GetDict(p.Resources.Pattern); found {
+			for _, key := range dict.Keys() {
+				resources[key] = dict.Get(key)
+			}
+		}
+	}
+	if p.Resources.Shading != nil {
+		if dict, found := core.GetDict(p.Resources.Shading); found {
+			for _, key := range dict.Keys() {
+				resources[key] = dict.Get(key)
+			}
+		}
+	}
+	if p.Resources.ProcSet != nil {
+		if dict, found := core.GetDict(p.Resources.ProcSet); found {
+			for _, key := range dict.Keys() {
+				resources[key] = dict.Get(key)
+			}
+		}
+	}
+	if p.Resources.Properties != nil {
+		if dict, found := core.GetDict(p.Resources.Properties); found {
+			for _, key := range dict.Keys() {
+				resources[key] = dict.Get(key)
+			}
+		}
+	}
+
+	return resources
+}
+
+// NewPdfAppender creates a new Pdf appender from a Pdf reader.
+func NewPdfAppender(reader *PdfReader) (*PdfAppender, error) {
+	a := &PdfAppender{}
+	a.rs = reader.rs
+	a.Reader = reader
+	a.parser = a.Reader.parser
+	if _, err := a.rs.Seek(0, io.SeekStart); err != nil {
+		return nil, err
+	}
+	var err error
+	a.roReader, err = NewPdfReader(a.rs)
+	if err != nil {
+		return nil, err
+	}
+	for _, idx := range a.Reader.GetObjectNums() {
+		if a.greatestObjNum < idx {
+			a.greatestObjNum = idx
+		}
+	}
+	a.xrefs = a.parser.GetXrefTable()
+	a.hasNewObject = make(map[core.PdfObject]struct{})
+	for _, p := range a.roReader.PageList {
+		a.pages = append(a.pages, p)
+	}
+	a.acroForm = a.roReader.AcroForm
+
+	return a, nil
+}
+
+func (a *PdfAppender) addNewObjects(obj core.PdfObject) {
+	if _, ok := a.hasNewObject[obj]; ok || obj == nil {
+		return
+	}
+	switch v := obj.(type) {
+	case *core.PdfIndirectObject:
+		if v.GetParser() != a.roReader.parser {
```

Comment:
Can you briefly describe the rationale for comparing the object's parser with the immutable reader parser in this function?  Would be good to get a brief summary of how the "change" detection works, for example in a comment above the function `addNewObjects`.